### ISSUE TITLE
Sample Code

### ISSUE THREAD

**Issue by pdgmdm** [xricht17/twpp#14]
Hi, would it be possible to add some sample projects for macos/Windows ? I would like to create a DS for a webcam...

**Comment by xricht17** [comment]
Hey, I'll see what I can do. Though, I am not sure it will help you that much. `TWPP` is meant to make your life easier, not to hide `TWAIN`, you actually have to understand the original `API`. This applies even more to `DS` than to `APP`, as `DS` is more challenging to implement correctly. Please see `TWAIN` specification before you attempt any serious work: https://www.twain.org/specification/.

**Comment by pdgmdm** [comment]
Hi,
I am struggling to setup the data source.
As an example, I want to add three capabilities: SupportedCaps, IXResolution and IYResolution. How should I do this:




        // is this correct ? how should it be done properly ??

        std::map<CapType,Capability*> caps;

        caps[CapType::SupportedCaps] = &Capability::createArray<CapType::SupportedCaps>(3);

        caps[CapType::SupportedCaps]->array<CapType::SupportedCaps>().at(0) = CapType::SupportedCaps;

        caps[CapType::SupportedCaps]->array<CapType::SupportedCaps>().at(1) = CapType::IXResolution;

        caps[CapType::SupportedCaps]->array<CapType::SupportedCaps>().at(2) = CapType::IXResolution;



        caps[CapType::IXResolution] = &Capability::createEnumeration<CapType::IXResolution>(1);

        caps[CapType::IXResolution]->enumeration<CapType::IXResolution>().at(0) = 640;



        caps[CapType::IYResolution] = &Capability::createEnumeration<CapType::IXResolution>(1);

        caps[CapType::IYResolution]->enumeration<CapType::IXResolution>().at(0) = 480;


Also, in the capbilityGet method: how should I return the caps ?



    virtual Result capabilityGet(const Identity& origin, Capability& data) override{

        // capability Msg::Get operation



        if (caps.find(data.type()) != caps.end())

        {

            logmsg << "found" << std::endl;

            // how should I fill in the data ??

            return success();

        }

        else

        {

            logmsg << "unsupported" << std::endl;

            return capUnsupported();

        }



        return success();
    }
-

**Comment by xricht17** [comment]
Hi, I have not forgotten about that sample code I promised you earlier, I hope to put something together next week.

---

About you question. I would not recommend to use these structures to hold your actual data, use them only to pass information within the `TWAIN` API. You should consider using them for other purposes only if you need to optimize your code. Otherwise, they will get into your way more often than not.

`Capability` class, and most other `TWPP` classes, are designed to be small containers that utilize C++11 move semantics, and to disallow copy semantics. That means you should never use dynamic memory to create them yourself (bad idea in C++ in general), nor should you use pointers to them.

---

In your example, you create a temporary object with `createArray`, and then store its address in a map:
> `caps[CapType::SupportedCaps] = &Capability::createArray<CapType::SupportedCaps>(3);`

This will not work. A new object is created in the `createArray` method, it is move-returned to the caller, the caller takes its address, and discards the object as it is not assigned anywhere. The destructor of the new object is run before you assign your, now useless, pointer to the map. Have a look at `std::unique_ptr` - `TWPP` classes should be handled in much the same way.

If you truly need a container to hold `Capability` objects, the correct way, in case of `std::map` could be:
```c++
Capability cap = Capability::createArray<CapType::SupportedCaps>({CapType::SupportedCaps, CapType::IXResolution, CapType::IXResolution});

std::map<CapType, Capability> caps;
caps.emplace(std::make_pair(CapType::SupportedCaps, std::move(cap)))
```

This creates a `Capability` instance with the 3 specified entries, sets it to variable `cap`, then creates a map between `CapType` and `Capability`, and finally moves `cap` into the map. The content of `cap` is now inside the map at index 0, and the `cap` itself is left empty.

Have a look at helper methods in `Capability` that allow you to create a fully initialized instance in one call, instead of having to assign members yourself as an additional step.

---

To return a `Capability`, or just anything except status, you _move_ it into the proper method parameter. In this case, it is the `data` parameter, that is both input and output parameter. This is the main reason you should not use `TWPP` to store your data. The data of `Capability` instance is _moved_ outside your domain, and you are left with an empty container. It is due to the fact that, in `TWAIN`, application is the endpoint that usually frees-up memory allocated by data source.

**Comment by xricht17** [comment]
Added a sample source in d7bae8b550b46ece7c089175940c34e121ab31cc

**Comment by pdgmdm** [comment]
Hi,
The sample application is working fine, however when I try to do some testing using apps like "Twister TWAIN Analyser" or "Inspector TWAIN" it does not work correctly.

For the "Inspector TWAIN" I get an SeqError when the app tries to close the DS when a transfer is started.
For the "Twister TWAIN Analyser" the app always crashes after many Get/GetCurrent/GetDefault calls.

Do you have tried this before ?

**Comment by xricht17** [comment]
There still seems to be an issue with old DSM support. I hoped it was resolved in f2227210c737f151298c6ffe5e2e61df5971e5ee. Obvously it was not. Created issue #15.

**Comment by xricht17** [comment]
Issue with Twister TWAIN Analyser seems to be due to the app passing uninitialized handle to the DS, thus resulting in an attempt to free invalid memory.

**Comment by xricht17** [comment]
Fixed the issue with Twister TWAIN Analyser in 5b2af944dd4630ef4993ecee8e7a6bff20783a80. Issue #15 will need more time. I am going to look at it at the weekend.

**Comment by xricht17** [comment]
Fixed the freeze in Inspector Twain (issue #15), the example DS now passes certification with 0 errors.
Twister TWAIN Analyser requires support for setting scan area for some reason. I do not think it is required by the `TWAIN` specification. Though, I will see what I can do in #17.
